AK/BigInts.h

```diff
@@ -0,0 +1,546 @@
+
+        if (low_carry && !high_carry)
+            high_carry = Checked<T>::addition_would_overflow(m_high + other.high(), 1u);
+
+        carry |= high_carry;
+
+        return {
+            m_low + other.low(),
+            m_high + other.high() + low_carry
+        };
+    }
+    template<Unsigned U>
+    requires(IsBaseOf<UBigInt, T>&& IsSame<U, R>) constexpr R&& addc(const U& other, bool& carry) const
+    {
+        bool low_carry = false;  //Checked<T>::addition_would_overflow(m_low, other.low());
+        bool high_carry = false; // Checked<T>::addition_would_overflow(m_high, other.high());
+
+        T lower = m_low.addc(other.low(), low_carry);
+        T higher = m_high.addc(other.high(), high_carry);
+
+        if (low_carry)
+            higher = higher.addc(1, high_carry);
+
+        carry |= high_carry;
+
+        return { lower, higher };
+    }
+    template<Unsigned U>
+    requires(sizeof(T) < sizeof(U)) constexpr R&& addc(const UBigInt<U>& other, bool& carry) const
+    {
+        return other.addc(*this, carry);
+    }
+
+    // FIXME:  subc for sizeof(U)>sizeof(T)
+    template<Unsigned U>
+    requires(sizeof(T) > sizeof(U)) constexpr R&& subc(const U& other, bool& carry) const
+    {
+        bool low_carry = m_low < other;
+
+        if (low_carry)
+            carry |= m_high == 0;
+
+        T lower = m_low - other;
+        T higher = m_high - low_carry;
+
+        return { lower, higher };
+    }
+    constexpr R&& subc(const R& other, bool& carry) const
+    {
+        bool low_carry = m_low < other;
+        bool high_carry = m_high < other.high();
+
+        if (low_carry && !high_carry)
+            high_carry = m_high == 0;
+
+        T lower = m_low - other.low();
+        T higher = m_high - other.high() - low_carry;
+
+        return { lower, higher };
+    }
+
+    template<Unsigned U>
+    constexpr R&& operator+(U other) const
+    {
+        bool carry; // unused
+        return addc(other, carry);
+    }
+    template<Unsigned U>
+    constexpr R&& operator-(U other) const
+    {
+        bool carry; // unused
+        return subc(other, carry);
+    }
+
+    template<Unsigned U>
+    constexpr R& operator+=(U other)
+    {
+        bool carry;
+        *this = addc(other, carry);
+        return *this;
+    }
+    template<Unsigned U>
+    constexpr R& operator-=(U other)
+    {
+        bool carry;
+        *this = subc(other, carry);
+        return *this;
+    }
+
+    constexpr R&& operator++()
+    {
+        // x++
+        auto old = *this;
+        *this += 1;
+        return old;
+    }
+    constexpr R&& operator++(int)
+    {
+        // ++x
+        *this += 1;
+        return *this;
+    }
+    constexpr R&& operator--()
+    {
+        // x--
+        auto old = *this;
+        *this -= 1;
+        return old;
+    }
+    constexpr R&& operator--(int)
+    {
+        // --x
+        *this -= 1;
+        return *this;
+    }
+
+    // FIXME: following maybe really inefficient
```

Comment:
I agree, division by subtraction shouldn't be the way to go here. But I can't come up with something better, I think even hardware dividers work similarly (especially if you need both the quotient and the remainder). Maybe ask the crypto people :^)

Reply:
It's sadly broken at the moment...